File: src/main/java/com/telegram/codex/integration/telegram/application/webhook/TelegramCommandResponder.java
package com.telegram.codex.integration.telegram.application.webhook;

import com.telegram.codex.conversation.application.ProcessedUpdateService;
import com.telegram.codex.conversation.application.session.SessionService;
import com.telegram.codex.integration.telegram.application.CompactResultSender;
import com.telegram.codex.integration.telegram.application.port.out.TelegramGateway;
import com.telegram.codex.integration.telegram.domain.InboundMessage;
import org.springframework.stereotype.Component;

import java.util.List;

@Component
public class TelegramCommandResponder {

    private final ProcessedUpdateService processedUpdateService;
    private final TelegramGateway telegramClient;
    private final CompactResultSender compactResultSender;

    public TelegramCommandResponder(
        ProcessedUpdateService processedUpdateService,
        TelegramGateway telegramClient,
        CompactResultSender compactResultSender
    ) {
        this.processedUpdateService = processedUpdateService;
        this.telegramClient = telegramClient;
        this.compactResultSender = compactResultSender;
    }

    public void reply(InboundMessage message, String text) {
        telegramClient.sendMessage(message.chatId(), text, List.of(), true);
        processedUpdateService.markProcessed(message);
    }

    public void sendCompactResult(InboundMessage message, SessionService.SessionCompactResult result) {
        compactResultSender.send(message.chatId(), result);
        processedUpdateService.markProcessed(message);
    }
}

File: src/telegram/commands/telegram_command_responder.rs
use std::sync::Arc;

use anyhow::Result;

use crate::conversation::reply::processed_update_service::ProcessedUpdateService;
use crate::conversation::session::session_service::SessionCompactResult;
use crate::telegram::commands::compact_result_sender::CompactResultSender;
use crate::telegram::shared::inbound_message::InboundMessage;
use crate::telegram::shared::telegram_types::TelegramGateway;

pub struct TelegramCommandResponder {
    processed_update_service: Arc<ProcessedUpdateService>,
    telegram_client: Arc<dyn TelegramGateway>,
    compact_result_sender: Arc<CompactResultSender>,
}

impl TelegramCommandResponder {
    pub fn new(processed_update_service: Arc<ProcessedUpdateService>, telegram_client: Arc<dyn TelegramGateway>, compact_result_sender: Arc<CompactResultSender>) -> Self {
        Self {
            processed_update_service,
            telegram_client,
            compact_result_sender,
        }
    }

    pub async fn reply(&self, message: &InboundMessage, text: &str) -> Result<()> {
        self.telegram_client.send_message(&message.chat_id, Some(text), &[], true).await?;
        self.processed_update_service.mark_processed_message(message).await
    }

    pub async fn send_compact_result(&self, message: &InboundMessage, result: &SessionCompactResult) -> Result<()> {
        self.compact_result_sender.send(&message.chat_id, result).await?;
        self.processed_update_service.mark_processed_message(message).await
    }
}
